ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,438 @@
+	 * <=> MAX_E = (divisor_high * B * (B^k - 1)) / (divisor_high * (divisor_high * B^k + B^k - 1))
+	 *
+	 * divisor_high uses the minimum value, but want to see how the number of digits affects the error, so represent
+	 * the minimum value as:
+	 * divisor_high = 10^x (any x ∈ ℕ)
+	 * Since B = 10^b, the formula becomes:
+	 * MAX_E = (divisor_high * B * (B^k - 1)) / (divisor_high * (divisor_high * B^k + B^k - 1))
+	 * <=> MAX_E = (10^x * 10^b * ((10^b)^k - 1)) / (10^x * (10^x * (10^b)^k + (10^b)^k - 1))
+	 *
+	 * Now let's make an approximation. Remove -1 from the numerator. Approximate the numerator to be
+	 * large and the denominator to be small, such that MAX_E is less than this expression.
+	 * Also, the denominator "(10^b)^k - 1" is never a negative value, so delete it.
+	 *
+	 * MAX_E = (10^x * 10^b * ((10^b)^k - 1)) / (10^x * (10^x * (10^b)^k + (10^b)^k - 1))
+	 * MAX_E < (10^x * 10^b * (10^b)^k) / (10^x * 10^x * (10^b)^k)
+	 * <=> MAX_E < 10^b / 10^x
+	 * <=> MAX_E < 10^(b - x)
+	 *
+	 * Therefore, found that if the number of digits in base B and divisor_high are equal, the error will be less
+	 * than 1 regardless of the number of digits in the value of k.
+	 *
+	 * Here, B is BC_VECTOR_BOUNDARY_NUM, so adjust the number of digits in high part of divisor to be BC_VECTOR_SIZE + 1.
+	 */
+	size_t divisor_top_digits = divisor_len % BC_VECTOR_SIZE;
+	if (divisor_top_digits == 0) {
+		divisor_top_digits = BC_VECTOR_SIZE;
+	}
+	size_t high_part_shift = POW_10_LUT[BC_VECTOR_SIZE - divisor_top_digits + 1];
+	size_t low_part_shift = POW_10_LUT[divisor_top_digits - 1];
+	BC_VECTOR divisor_high_part = divisor_vectors[divisor_top_index] * high_part_shift + divisor_vectors[divisor_top_index - 1] / low_part_shift;
+	for (size_t i = 0; i < quot_arr_size; i++) {
+		BC_VECTOR numerator_high_part = numerator_vectors[numerator_top_index - i] * high_part_shift + numerator_vectors[numerator_top_index - i - 1] / low_part_shift;
+
+		/*
+		 * Determine the temporary quotient.
+		 * The maximum value of numerator_high is divisor_high * B in the previous example, so here numerator_high_part is
+		 * divisor_high_part * BC_VECTOR_BOUNDARY_NUM.
+		 * The number of digits in divisor_high_part is BC_VECTOR_SIZE + 1, so even if divisor_high_part is at its maximum value,
+		 * it will never overflow here.
+		 */
+		numerator_high_part += div_carry * BC_VECTOR_BOUNDARY_NUM * high_part_shift;
+
+		/* numerator_high_part < divisor_high_part => quotient == 0 */
+		if (numerator_high_part < divisor_high_part) {
+			quot_vectors[quot_top_index - i] = 0;
+			div_carry = numerator_vectors[numerator_top_index - i];
+			numerator_vectors[numerator_top_index - i] = 0;
+			continue;
+		}
+
+		BC_VECTOR quot_guess = numerator_high_part / divisor_high_part;
+
+		/* For sub, add the remainder to the high-order digit */
+		numerator_vectors[numerator_top_index - i] += div_carry * BC_VECTOR_BOUNDARY_NUM;
```

Comment:
The "for sub" here is throwing me off, what are you trying to say precisely?

Reply:
`numerator_high_part` is only used to calculate the quotient, and `numerator_vectors` does not contain `div_carry` here yet.

After this, add `div_carry` to `numerator_vectors` for `numerator_vectors - (divisor_vectors * quot_guess)`.